### PR TITLE
Add PatchMapWithPatchingMove

### PR DESCRIPTION
**Pull request by Ericson2314**
From https://github.com/reflex-frp/reflex/pull/372

CC @ryantrinkle

### CODE REVIEW

**Comment by Ericson2314** [comment]
https://travis-ci.org/reflex-frp/patch/builds/641036352?utm_source=github_status&utm_medium=notification is CIing the wrong PR. WTF!

File: src/Data/Patch/MapWithPatchingMove.hs [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+  -> Map k (PatchTarget p) -> PatchMapWithPatchingMove k p
+patchThatSortsMapWith cmp m = PatchMapWithPatchingMove $ Map.fromList $ catMaybes $ zipWith g unsorted sorted
+  where unsorted = Map.toList m
+        sorted = sortBy (cmp `on` snd) unsorted
+        f (to, _) (from, _) = if to == from then Nothing else
+          Just (from, to)
+        reverseMapping = Map.fromList $ catMaybes $ zipWith f unsorted sorted
+        g (to, _) (from, _) = if to == from then Nothing else
+          let Just movingTo = Map.lookup from reverseMapping
+          in Just (to, NodeInfo (From_Move from mempty) $ Just movingTo)
```
Comment (xplat) [inline src/Data/Patch/MapWithPatchingMove.hs:237]: where does this generate the "To" nodes?  afaict it doesn't ...

Comment (xplat) [reply src/Data/Patch/MapWithPatchingMove.hs:237]: nevermind, i'm blind; it would be harder to miss this if To were not a type synonym tho ...

**Review by xplat — APPROVED** [review]

File: src/Data/Patch/MapWithPatchingMove.hs (hunk 1 above)
Comment (Ericson2314) [reply src/Data/Patch/MapWithPatchingMove.hs:237]: Yes I had the same thought, and made it its own datatype in the other PR.

File: src/Data/Monoid/DecidablyEmpty.hs [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+  isEmpty = Map.null
+instance DecidablyEmpty (Seq.Seq v) where
+  isEmpty = Seq.null
+instance Ord k => DecidablyEmpty (Set.Set k) where
+  isEmpty = Set.null
+
+-- dependent-map
+
+instance GCompare k => DecidablyEmpty (DMap.DMap k v) where
+  isEmpty = DMap.null
```
Comment (Ericson2314) [inline src/Data/Monoid/DecidablyEmpty.hs:115]: @danbornside points out these could use `Foldable`'s `null`; and that might be the better default method too.

(At the very least, we can provide newtypes for both for `deriving via`.)

**Comment by Ericson2314** [comment]
The other failing travis was testing the wrong commit!

**Pull request merged**